### ISSUE TITLE
Migrate from package-relative imports to `wp-calypso-client` imports

### ISSUE THREAD

**Issue by scinos** [Automattic/wp-calypso#44602]
### Problem

Calypso allows package-relative imports, like

```js
import config from 'config';
import userFactory from 'lib/user';
```

While this saves us from having to type an endless stream of `../../../../../` for relative imports, it has some drawbacks: it makes the code harder to statically analyze, and such relative import has to be configured in all our tools that do module resolution (node.js, webpack, typescript, IDEs...).

### Solution

We will migrate to `wp-calypso-client` imports:

```js
import config from 'wp-calypso-client/config';
import userFactory from 'wp-calypso-client/lib/user';
```

This works because `./client` is actually an NPM package called `wp-calypso-client`, and it is referenced as a dependency in the root's `package.json`. 

### Migration plan

We will use eslint rules to do the transition:

- [x] Rename the package in `./client/` (currently named `wp-calypso`) to `wp-calypso-client` (#44637)
- [x] Write a new rule `wpcalypso/no-package-relative-imports` to detect the old pattern and auto-fix it to the new pattern. (#44479, #44557)
- [x] Change our config to support package-relative and `wp-calypso-client` imports at the same time (#44600)
- [x] Automatically convert all files.
- [ ] When we are confident there are no open branches with the old system, delete the config that supports package-relative imports.

**Comment by ockham** [comment]
> This works because `./client` is actually an NPM package called `wp-calypso`, and it is referenced as a dependency in the root's `package.json`. 

This could turn out a bit confusing to people (myself included :sweat_smile: ), since it's not totally obvious that in an import path like

`import userFactory from 'wp-calypso/lib/user';`

the `wp-calypso` part actually translates to `client/`. (People might go looking for a `wp-calypso/` folder instead. Eventually, we'll accept it as another Calypso-ism, but... do we really _want_ another Calypso-ism? :grimacing: )

I'm much in favor of making that path explicit though! How about

1. We rename the `client/` path to `wp-calypso/`, or both the path and package name to just `calypso`?
2. We do 1., and additionally move `wp-calypso/` under `apps/`? :grimacing:

**Comment by scinos** [comment]
I'm ok with both suggestions, but I do think we need to them step by step to avoid overcomplicating and already quite complicated PR.

Once we have removed all package-relative imports, it should be easier to rename the package `wp-calypso` to `wp-calypso-client` (or whatever name we decide). Renaming (or moving) `./client/` to something else would be quite challenging, that path may be references in a ton of places. I do think we should eventually do it, but not necessarily linked to the change proposed in this PR.

**Comment by ockham** [comment]
> Do you think it is worth renaming the package first, and then doing the replacement of package-relative imports to the new package name?

I think it might be worth considering :thinking: Since the package name isn't currently used explicitly in imports, it might be less work to change it _before_ making it explicit.

**Comment by jsnajdr** [comment]
I agree that we should rename `wp-calypso-client` to something nicer: `wp-calypso` or even just `calypso` (my favorite).

The `wp-calypso` name is currently used by the root `package.json` (at the top of the monorepo), but that name can probably be outright removed. Jest also doesn't specify a package name in their root [package.json](https://github.com/facebook/jest/blob/master/package.json) and it seems to work OK for them.

If we rename the package first and only then change all the imports, it means only one monster diff instead of two. That's why I'd prefer to agree on the new name first, but I don't feel strongly about it otherwise.

I don't know what to think about renaming the physical `client/` folder. I'm afraid the rename will break many existing links and references.

**Comment by simison** [comment]
> The wp-calypso name is currently used by the root package.json (at the top of the monorepo), but that name can probably be outright removed. Jest also doesn't specify a package name in their root package.json and it seems to work OK for them.

Would that affect anyone directly relying on pulling whole Calypso as a dependency? I remember there was some project doing that.

**Comment by scinos** [comment]
Current situation is:

- Root project is called `wp-calypso`. I don't see a good reason for not having a name there. I'd argue that we should rename it to something more generic because there other apps inside, not only Calypso. But that's another topic.

- `./client` is called `wp-calypso-client`. After typing it for a while I agree it feels too verbose. I don't have any strong opinion on the new name. I'd say we should use the word `client` somewhere in case we want to have `calypso-desktop`, `calypso-server`, etc packages in the future. But we can as well do the rename if/when we create those packages.

- We are already using `wp-calypso-client` in a few places in the code base. I'd say we continue to use `wp-calypso-client` for now in the preparation PRs (work that needs to be done before the mass rename), then rename a few instances to the new name and then do the mass rename.

TLDR: I don't mind renaming it to `calypso` or something else, but I'd like to continue using `wp-calypso-client` for the preparation PRs to reduce noise.

**Comment by jsnajdr** [comment]
> Would that affect anyone directly relying on pulling whole Calypso as a dependency? I remember there was some project doing that.

[WooCommerce Services](https://github.com/Automattic/woocommerce-services) used to declare `wp-calypso` as a `package.json` dependency (pointing to a GitHub URL), but that stopped working when we switched to a Lerna-managed monorepo. They since switched to a Git submodule more than a year ago. (The long-term solution still being modularizing all used Calypso components and then consuming individual NPM packages).

I don't insist that we remove the name, just pointing out that it's even possible (it was a surprise to me), as the Jest monorepo shows. Another option is `wp-calypso-monorepo`. Lerna, for example, uses this `-monorepo` suffix to name the root package.

**Comment by jsnajdr** [comment]
> I don't see a good reason for not having a name there.

The good reason is that the name doesn't need to be there. In other words, very good reasons are needed for _adding_ things, not for _removing_ them 🙂 Naming the root package `wp-calypso-monorepo` is another good option.

> I'd argue that we should rename it to something more generic because there other apps inside, not only Calypso. But that's another topic.

If there is an app in the monorepo that doesn't have nothing to do with Calypso, either the app or the framework, then it probably shouldn't be there. Adding the Editing Toolkit (aka FSE) was probably a mistake.

If there is no support for changing the `wp-calypso-client` name now, I'm OK with keeping it. Just keep in mind that every mass rename like this will invalidate most open PRs, namely these that touch the imports section in any way.

**Comment by tyxla** [comment]
> > I don't see a good reason for not having a name there.
> 
> The good reason is that the name doesn't need to be there. In other words, very good reasons are needed for _adding_ things, not for _removing_ them 🙂 Naming the root package `wp-calypso-monorepo` is another good option.

I'd be surprised if anyone would want to use the entire calypso monorepo as a package these days, so agree that a name isn't necessarily needed there. A name could be useful solely so we know what we're referencing when we're talking about it.

While I don't strongly object renaming it to `wp-calypso-monorepo`, I think keeping it `wp-calypso` to reflect the repo name makes the most sense. I've also seen a recurring pattern for naming such monorepos `@organization/repo` too, so in that case, that'd be `@automattic/wp-calypso`, just like Gutenberg packages are named. However, if we follow the Gutenberg suit, we'll also notice that the root package is called just `gutenberg` - another reason to just keep it `wp-calypso`.

> `./client` is called `wp-calypso-client`. After typing it for a while I agree it feels too verbose. I don't have any strong opinion on the new name. I'd say we should use the word `client` somewhere in case we want to have `calypso-desktop`, `calypso-server`, etc packages in the future. But we can as well do the rename if/when we create those packages.

I generally agree about the rationale to use `wp-calypso-client` there. Let's do that.

Something I prefer though is that in imports we use `import from 'client/components/accordion'` instead of `import from 'wp-calypso-client/components/accordion'`. I'm sure I'm not alone in the fact that I've grown too attached to relying on webpack module resolution to just work inside the "client" directory without specifically including it in the import path. 

> I don't mind renaming it to calypso or something else, but I'd like to continue using wp-calypso-client for the preparation PRs to reduce noise.

Totally agree to do it incrementally and step by step to reduce noise.

**Comment by scinos** [comment]
> > I don't see a good reason for not having a name there.

> The good reason is that the name doesn't need to be there. In other words, very good reasons are needed for adding things, not for removing them 🙂

The name is _already_ there. Can anyone think of one scenario where having a name in an unpublished package is problematic at all? What problem are we trying to solve/avoid by having a nameless root package? Because I can imagine how there may be static analysis tools out there that assume (incorrectly, agreed) that all package.json have a name and may break.

>  I'm sure I'm not alone in the fact that I've grown too attached to relying on webpack module resolution to just work inside the "client" directory without specifically including it in the import path.

I hope that the eslint rule (with auto-fix!) will change help here.

> Just keep in mind that every mass rename like this will invalidate most open PRs, namely these that touch the imports section in any way.

That can be mitigated if we support both names for a while (which is the plan for this change, support package-relative _and_ package imports). I agree it won't be a trivial migration, but maybe we should focus on making that migration easier because we may need to change the package name again. 

Let's use the learnings of this migration to avoid spreading the source of truth for the package name, webpack aliases and other supporting configs across the whole repo to make a future migration easier.

**Comment by tyxla** [comment]
Ditto. Let's keep both for a while, and try to be vocal to everyone about what's the recommended way and why we're doing it. We could also post a PSA for folks to rebase their PRs and change to absolute imports, and manually go through some PRs to ask folks to rebase/re-review them again with the changes. And as we discussed in Slack, we could also try to automate a check for enforcing a rebase of PRs with older `HEAD`.

**Comment by jsnajdr** [comment]
> The name is already there. Can anyone think of one scenario where having a name in an unpublished package is problematic at all?

The only problematic scenario is if we wanted to use the `wp-calypso` name for the app/package in `client/`, instead of the longer `wp-calypso-client`. Then the fact that the name is already "taken" by the root package would be an issue.
